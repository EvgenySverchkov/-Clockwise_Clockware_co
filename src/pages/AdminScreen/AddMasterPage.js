import React, { useEffect } from "react";
import PropTypes from "prop-types";
import { useDispatch, useSelector } from "react-redux";

import postData from "../../components/AdminScreen/services/postData";
import SubscribeBtn from "../../components/FormComponents/Button";

import FormGroup from "../../components/FormComponents/FormGroup";
import NameField from "../../components/FormComponents/TextField";
import NumField from "../../components/FormComponents/NumField";
import Label from "../../components/FormComponents/Label";

import SelectTownsField from "../../components/CompleteFormFields/SelectTownsField";

import { SERVERDOMAIN } from "../../services/serverUrls";

import { addNewMaster } from "../../store/adminPanel/masters/actions";
import {
  changeAddNewMasterFormIsLoad,
  changeSuccessModalDataAdmin,
  changeModalWarningDataAdmin,
} from "../../store/adminPanel/services/actions";
import { townsInit } from "../../store/adminPanel/towns/actions";

function AddMasterPage({ history }) {
  const state = useSelector((state) => {
    return {
      newMasterFormIsLoad: state.mainAdminPanelReduser.newMasterFormIsLoad,
      townsArr: state.townReduser.towns,
    };
  });
  const dispatch = useDispatch();

  useEffect(function () {
    getTownsFromServerToState();
  }, []);

  function getTownsFromServerToState() {
    const headers = {
      Authorization: "Bearer " + sessionStorage.getItem("token") || "",
      "Content-Type": "application/json",
    };
    fetch(`${SERVERDOMAIN}/towns`, { headers })
      .then((json) => json.json())
      .then((data) => dispatch(townsInit(data)));
  }
  function handler(e) {
    e.preventDefault();
    let name = e.target.name.value;
    let rating = e.target.rating.value;
    let towns = selectCheckedTowns(e.target.elements);
    dispatch(changeAddNewMasterFormIsLoad(true));
    let infoObj = {
      name,
      rating,
      towns: towns.join(","),
    };
    postData(`${SERVERDOMAIN}/masters/post`, infoObj)
      .then((data) => {
        dispatch(changeAddNewMasterFormIsLoad(false));
        if (data.success) {
          dispatch(addNewMaster(data.payload));
          dispatch(
            changeSuccessModalDataAdmin({
              msg: data.msg,
              backBtnTxt: "Back to masters list",
              backTo: "/admin/mastersList",
            })
          );
          history.push("/admin");
          document.getElementById("callSuccessModalBtn").click();
        } else {
          dispatch(changeModalWarningDataAdmin({ msg: data.msg }));
          document.getElementById("callWarningModalBtn").click();
        }
      })
      .catch((err) => alert(err));

    function selectCheckedTowns(elements) {
      let newArr = Array.from(elements);
      let towns = [];
      newArr.forEach((item) => {
        if (item.className.match(/\btowns\b/)) {
          if (item.checked) {
            towns.push(item.value);
          }
        }
      });
      return towns;
    }
  }
  return (
    <form onSubmit={handler} className="mt-4 row justify-content-center">
      <FormGroup>
        <Label forId={"rating"}>Enter rating</Label>
        <NumField max={5} min={0} id={"rating"} />
      </FormGroup>
      <FormGroup>
        <Label forId={"name"}>Enter name</Label>
        <NameField id={"name"} name={"name"} />
      </FormGroup>
      <SelectTownsField townsArr = {state.townsArr}/>
      <SubscribeBtn isLoad={state.newMasterFormIsLoad} value={"Add"} />
    </form>
  );
}

AddMasterPage.propTypes = {
  history: PropTypes.object,
};

export default AddMasterPage;
