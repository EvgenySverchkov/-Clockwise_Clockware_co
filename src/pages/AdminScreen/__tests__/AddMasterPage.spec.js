import React from "react";
import {create, act} from "react-test-renderer";
import {Provider} from "react-redux";
import configMockStore from "redux-mock-store";

import Context from "../../../ContextComponent";
import AddMasterPage from "../AddMasterPage";

const mockStore = configMockStore();

describe("Test of <AddMasterPage />", ()=>{
    global.fetch = jest.fn(()=>Promise.resolve({json: jest.fn(()=>Promise.resolve())}));
    const initialStore = {
        masterReducer: {
            newMasterFormIsLoad: false
        },
        townReduser: {
            towns: [{id:1, name: "Name1"}, {id:2, name: "Name2"}]
        },
    };
    const mockContextValue = {
        openErrorWindowWithMsg: jest.fn(),
        openWarningTooltip: jest.fn(),
        warningTooltipMsg: "Message"
    };
    it("Default view", async()=>{
        const store = mockStore(initialStore);
        let component;
        await act(async ()=>{
            component = await create(
                <Provider store = {store}>
                    <Context.Provider value = {mockContextValue}>
                        <AddMasterPage />
                    </Context.Provider>
                </Provider>
            );
        });
        expect(component.toJSON()).toMatchSnapshot();
    });
    it("The view when towns array is empty", async()=>{
        const store = mockStore({...initialStore, townReduser: {towns: []}});
        let component;
        await act(async ()=>{
            component = await create(
                <Provider store = {store}>
                    <Context.Provider value = {mockContextValue}>
                        <AddMasterPage />
                    </Context.Provider>
                </Provider>
            );
        });
        expect(component.toJSON()).toMatchSnapshot();
    });

    it("The view after click on submit button and waiting response from server", async()=>{
        const store = mockStore({...initialStore, masterReducer: {newMasterFormIsLoad: true}});
        let component;
        await act(async ()=>{
            component = await create(
                <Provider store = {store}>
                    <Context.Provider value = {mockContextValue}>
                        <AddMasterPage />
                    </Context.Provider>
                </Provider>
            );
        });
        expect(component.toJSON()).toMatchSnapshot();
    });
});
