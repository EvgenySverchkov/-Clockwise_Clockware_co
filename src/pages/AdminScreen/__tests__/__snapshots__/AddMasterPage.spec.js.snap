// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test of <AddMasterPage /> Default view 1`] = `
<form
  className="mt-4 row justify-content-center"
  onBlur={[Function]}
  onSubmit={[Function]}
>
  <div
    className="form-group row text-center text-sm-left col-12 col-md-10 col-lg-8"
  >
    <label
      className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label font-weight-bold"
      htmlFor="rating"
    >
      Enter rating
    </label>
    <div
      className="col-sm-3"
    >
      <input
        aria-describedby={null}
        className="form-control"
        id="rating"
        max={5}
        min={0}
        onBlur={[Function]}
        onChange={null}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        title="Message"
        type="number"
      />
    </div>
  </div>
  <div
    className="form-group row text-center text-sm-left col-12 col-md-10 col-lg-8"
  >
    <label
      className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label font-weight-bold"
      htmlFor="name"
    >
      Enter name
    </label>
    <div
      className="col-sm-8 col-md-8"
    >
      <input
        aria-describedby={null}
        className="form-control"
        id="name"
        name="name"
        onBlur={[Function]}
        onChange={null}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        required={true}
        title="Message"
        type="text"
      />
    </div>
  </div>
  <div
    className="form-group  text-center text-sm-left col-12 col-md-10 col-lg-8"
  >
    <div
      className="mb-2 font-weight-bold"
    >
      Choose towns
    </div>
    <div>
      <div
        className="form-check-inline"
      >
        <label
          className="form-check-label"
          htmlFor="Name1"
        >
          <input
            className="form-check-input towns"
            id="Name1"
            name="towns"
            type="checkbox"
            value="Name1"
          />
          Name1
        </label>
      </div>
      <div
        className="form-check-inline"
      >
        <label
          className="form-check-label"
          htmlFor="Name2"
        >
          <input
            className="form-check-input towns"
            id="Name2"
            name="towns"
            type="checkbox"
            value="Name2"
          />
          Name2
        </label>
      </div>
    </div>
  </div>
  <div
    className="row justify-content-sm-center col-12"
  >
    <input
      className="btn btn-primary col-12 col-sm-4 mt-3"
      type="submit"
      value="Add"
    />
  </div>
</form>
`;

exports[`Test of <AddMasterPage /> The view after click on submit button and waiting response from server 1`] = `
<form
  className="mt-4 row justify-content-center"
  onBlur={[Function]}
  onSubmit={[Function]}
>
  <div
    className="form-group row text-center text-sm-left col-12 col-md-10 col-lg-8"
  >
    <label
      className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label font-weight-bold"
      htmlFor="rating"
    >
      Enter rating
    </label>
    <div
      className="col-sm-3"
    >
      <input
        aria-describedby={null}
        className="form-control"
        id="rating"
        max={5}
        min={0}
        onBlur={[Function]}
        onChange={null}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        title="Message"
        type="number"
      />
    </div>
  </div>
  <div
    className="form-group row text-center text-sm-left col-12 col-md-10 col-lg-8"
  >
    <label
      className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label font-weight-bold"
      htmlFor="name"
    >
      Enter name
    </label>
    <div
      className="col-sm-8 col-md-8"
    >
      <input
        aria-describedby={null}
        className="form-control"
        id="name"
        name="name"
        onBlur={[Function]}
        onChange={null}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        required={true}
        title="Message"
        type="text"
      />
    </div>
  </div>
  <div
    className="form-group  text-center text-sm-left col-12 col-md-10 col-lg-8"
  >
    <div
      className="mb-2 font-weight-bold"
    >
      Choose towns
    </div>
    <div>
      <div
        className="form-check-inline"
      >
        <label
          className="form-check-label"
          htmlFor="Name1"
        >
          <input
            className="form-check-input towns"
            id="Name1"
            name="towns"
            type="checkbox"
            value="Name1"
          />
          Name1
        </label>
      </div>
      <div
        className="form-check-inline"
      >
        <label
          className="form-check-label"
          htmlFor="Name2"
        >
          <input
            className="form-check-input towns"
            id="Name2"
            name="towns"
            type="checkbox"
            value="Name2"
          />
          Name2
        </label>
      </div>
    </div>
  </div>
  <div
    className="row justify-content-sm-center col-12"
  >
    <input
      className="btn btn-primary col-12 col-sm-4 mt-3"
      type="submit"
      value="Loading..."
    />
  </div>
</form>
`;

exports[`Test of <AddMasterPage /> The view when towns array is empty 1`] = `
<form
  className="mt-4 row justify-content-center"
  onBlur={[Function]}
  onSubmit={[Function]}
>
  <div
    className="form-group row text-center text-sm-left col-12 col-md-10 col-lg-8"
  >
    <label
      className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label font-weight-bold"
      htmlFor="rating"
    >
      Enter rating
    </label>
    <div
      className="col-sm-3"
    >
      <input
        aria-describedby={null}
        className="form-control"
        id="rating"
        max={5}
        min={0}
        onBlur={[Function]}
        onChange={null}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        title="Message"
        type="number"
      />
    </div>
  </div>
  <div
    className="form-group row text-center text-sm-left col-12 col-md-10 col-lg-8"
  >
    <label
      className="pl-0 pr-0 col-sm-4 col-md-4 col-form-label font-weight-bold"
      htmlFor="name"
    >
      Enter name
    </label>
    <div
      className="col-sm-8 col-md-8"
    >
      <input
        aria-describedby={null}
        className="form-control"
        id="name"
        name="name"
        onBlur={[Function]}
        onChange={null}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        required={true}
        title="Message"
        type="text"
      />
    </div>
  </div>
  <div
    className="form-group  text-center text-sm-left col-12 col-md-10 col-lg-8"
  >
    <div
      className="mb-2 font-weight-bold"
    >
      Choose towns
    </div>
    <div>
      -- List is empty --
    </div>
  </div>
  <div
    className="row justify-content-sm-center col-12"
  >
    <input
      className="btn btn-primary col-12 col-sm-4 mt-3"
      type="submit"
      value="Add"
    />
  </div>
</form>
`;
